_posts/2025-04-23-#31444.md

```diff
@@ -10,17 +10,95 @@ status: upcoming
+  feerate and individual feerate). This also linearizes transactions in an approximate order of least to most likely
+to be mined, but the linearization is not exactly opposite to the result of ancestor set-based sorting. This asymmetry
+is problematic:
+    - What if the first transaction selected for eviction is also the transaction that would be selected first for
+      mining? This cache eviction strategy is throwing away the most useful piece of data.
+    - Similarly, what if we have transactions that are "junk" (e.g. fall below the node's configured `-blockmintxfee`
+      and would thus never be selected for mining) but can't be kicked out because they have a high descendant score?
+    This mempool limitation necessitated the requirement that package transaactions be above the minimum relay feerate
+  (see [PR #26933](https://github.com/bitcoin/bitcoin/pull/26933)).
+- Due to similar limitations, Replace-by-Fee code cannot properly determine whether the proposed replacement(s) are, so
+  it uses approximations like individual feerate comparisons. These imperfect heuristics lead to pinning problems and
+fee-bumping inefficiencies.
 
+### TxGraph
+- The [main motivations](https://github.com/bitcoin/bitcoin/issues/27677#eviction-mining-not-opposite) for cluster mempool
+  are to address these problems.
+- A cluster mempool essentially keeps the entire mempool linearized at all times (lazily using `TxGraph`), which
+  makes it easy to quickly determine a transactions linearization position (including what the highest and lowest
+transactions are) and compare the current mempool with a potential one.
+[PR #31444](https://github.com/bitcoin/bitcoin/pull/31444) adds `TxGraph` functionality to assist in selecting
+transactions for block building and eviction:
+    - `TxGraph` adds a chunk index, a total ordering of all chunks in the graph across all clusters. A
+      [`ChunkOrder`
+comparator](https://github.com/bitcoin-core-review-club/bitcoin/commit/3429e9d79df1336cf1d0a61cb5f9bf028aa860b2#diff-c25ebcc6e4dc9c9bb21462f796fbb470cee3f0ff1e77bfc4033f4e8460f4482cR282)
+defines the order in which chunks would be mined.
+    - `BlockBuilder` can just iterate through the chunks in order. Also, since a total orderering of all graph entries
+      exists, we can get the “least likely to be mined” transaction without building a full mempool worth of block
+templates: simply take the last chunk in the index.
+    - `TxGraph` exposes a `BlockBuilder` to iterate over the chunks from best to worst. `BlockBuilder` can `Skip` chunks (i.e.
+      if they are too large to fit in the remaining block space), and those chunks' descendants will be subsequently
+skipped as well.
+    - `TxGraph` also exposes a `GetWorstMainChunk` method to identify the transactions that would be included in a block last.
+- The PR also adds a `GetMainStagingDiagrams` method to make it possible to compare the feerate diagram of the `TxGraph`
+  with and without its staged changes.
```

Comment:
The PR states:

> `TxGraph::GetMainStagingDiagrams()`, a function to obtain feerate diagrams for both the main graph and the staged changes to it, including only the clusters that differ between the two.

I'm not really sure what a feerate diagram is and why it's useful/who uses it. Since it seems to be the focus of the PR, adding some context on that could be useful?

(`TxGraph::GetBlockBuilder()` and `TxGraph::GetWorstMainChunk()` are much more intuitive to me.)

Reply:
Added a bit more background. I plan to emphasize this part less, as imo the block builder and eviction commits are the more novel / interesting parts of the PR.